File: CustomBehavior/app/src/main/java/com/app/behavior/behavior/helper/ViewOffsetBehavior.java
package com.app.behavior.behavior.helper;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;

import androidx.coordinatorlayout.widget.CoordinatorLayout;

/**
 * 封装了对水平和垂直方向偏移
 * ViewOffsetBehavior 的作用是方便改变控件的位置和获取偏移量
 * @param <V>
 */
public class ViewOffsetBehavior<V extends View> extends CoordinatorLayout.Behavior<V> {
//    当我们考虑一个滑动交互时，不要把滑动看做一个连续过程，
//    而要拆分成多个单独的循环过程，连续的滑动只不过是单独循环过程
//    在时间上不断重复而已；而滑动的单个循环过程，说到底都是对View进行偏移处理。
//    当看到一个复杂交互效果的时候，要学会拆分，一个是刚说的时间上拆分，
//    另一个方面就是要能拆分成多个单独效果的合成

    private ViewOffsetHelper mViewOffsetHelper;

    private int mTempTopBottomOffset = 0;
    private int mTempLeftRightOffset = 0;

    public ViewOffsetBehavior() {
    }

    public ViewOffsetBehavior(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    /**
     * 主要是创建ViewOffsetHelper、获取真实偏移量并且将child偏移到正确位置。
     */
    @Override
    public boolean onLayoutChild(CoordinatorLayout parent, V child, int layoutDirection) {
        // First let lay the child out
        layoutChild(parent, child, layoutDirection);

        if (mViewOffsetHelper == null) {
            mViewOffsetHelper = new ViewOffsetHelper(child);
        }
        mViewOffsetHelper.onViewLayout();

        if (mTempTopBottomOffset != 0) {
            mViewOffsetHelper.setTopAndBottomOffset(mTempTopBottomOffset);
            mTempTopBottomOffset = 0;
        }
        if (mTempLeftRightOffset != 0) {
            mViewOffsetHelper.setLeftAndRightOffset(mTempLeftRightOffset);
            mTempLeftRightOffset = 0;
        }

        return true;
    }

    protected void layoutChild(CoordinatorLayout parent, V child, int layoutDirection) {
        // Let the parent lay it out by default
        parent.onLayoutChild(child, layoutDirection);
    }

    public boolean setTopAndBottomOffset(int offset) {
        if (mViewOffsetHelper != null) {
            return mViewOffsetHelper.setTopAndBottomOffset(offset);
        } else {
            mTempTopBottomOffset = offset;
        }
        return false;
    }

    public boolean setLeftAndRightOffset(int offset) {
        if (mViewOffsetHelper != null) {
            return mViewOffsetHelper.setLeftAndRightOffset(offset);
        } else {
            mTempLeftRightOffset = offset;
        }
        return false;
    }

    public int getTopAndBottomOffset() {
        return mViewOffsetHelper != null ? mViewOffsetHelper.getTopAndBottomOffset() : 0;
    }

    public int getLeftAndRightOffset() {
        return mViewOffsetHelper != null ? mViewOffsetHelper.getLeftAndRightOffset() : 0;
    }
}

File: NestedScrollProject/nestedlibrary/src/main/java/com/moj/nestedlibrary/ViewOffsetBehavior.kt
package com.moj.nestedlibrary

import android.content.Context
import android.support.design.widget.CoordinatorLayout
import android.util.AttributeSet
import android.view.View

/**
 * @author : moj
 * @date : 2019/8/9
 * @description : 布局行为, 如何布局呢？
 *
 * 规则：比如现在有一个View称之为a， 另外一个View称之为b，我想b在a的右边。
 * 1.（b依赖a,所以需要为b设置behavior-解释为b依赖a）并设置layout_anchor为a-解释为b绑定a
 * 明确了布局要求之后，我们知道b的布局需要依赖a的位置，所以我们可以把ViewOffsetBehavior添加在b上，
 * 并设置layout_anchor为a，这可以理解为b绑定了a，
 * 此时我们设置b的layout_gravity属性，这个属性表明a与b的对齐关系，如设置left|top，表示a与b的左边和上边对齐，
 * 然后设置b的layout_anchorGravity属性，表明b与a的位置，left表示b在a的左边。
 *
 * ps：如果不设置layout_gravity属性，只设置layout_anchorGravity，那么结果是，如果是left，会是b的right边在a的中间，即b
 * 向右平移的a的一半宽度
 *
 * ps: 使用此种布局方式时margin属性无效，如需添加间隔，可以使用Space实现。
 */
open class ViewOffsetBehavior<V : View> : CoordinatorLayout.Behavior<V> {

    private var viewOffsetHelper: ViewOffsetHelper? = null
    private var tempTopBottomOffset = 0
    private var tempLeftRightOffset = 0

    val topAndBottomOffset: Int
        get() = if (this.viewOffsetHelper != null) this.viewOffsetHelper!!.getTopAndBottomOffset() else 0

    val leftAndRightOffset: Int
        get() = if (this.viewOffsetHelper != null) this.viewOffsetHelper!!.getLeftAndRightOffset() else 0

    constructor() {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    override fun onLayoutChild(parent: CoordinatorLayout, child: V, layoutDirection: Int): Boolean {
        this.layoutChild(parent, child, layoutDirection)
        if (this.viewOffsetHelper == null) {
            this.viewOffsetHelper = ViewOffsetHelper(child)
        }

        this.viewOffsetHelper!!.onViewLayout()
        if (this.tempTopBottomOffset != 0) {
            this.viewOffsetHelper!!.setTopAndBottomOffset(this.tempTopBottomOffset)
            this.tempTopBottomOffset = 0
        }

        if (this.tempLeftRightOffset != 0) {
            this.viewOffsetHelper!!.setLeftAndRightOffset(this.tempLeftRightOffset)
            this.tempLeftRightOffset = 0
        }

        return true
    }

    protected fun layoutChild(parent: CoordinatorLayout, child: V, layoutDirection: Int) {
        parent.onLayoutChild(child, layoutDirection)
    }

    fun setTopAndBottomOffset(offset: Int): Boolean {
        if (this.viewOffsetHelper != null) {
            return this.viewOffsetHelper!!.setTopAndBottomOffset(offset)
        } else {
            this.tempTopBottomOffset = offset
            return false
        }
    }

    fun setLeftAndRightOffset(offset: Int): Boolean {
        if (this.viewOffsetHelper != null) {
            return this.viewOffsetHelper!!.setLeftAndRightOffset(offset)
        } else {
            this.tempLeftRightOffset = offset
            return false
        }
    }

}
